(WIP) Create code-studio package

islemaster opened the pull request:
Introduces a _code-studio_ package (and associated pipeline) that should be used to build dashboard-specific JS assets.  This is an early step in a longer process of standardizing our static asset pipelines, and maybe making it possible to serve those assets via CDN.

This initial pass only moves the blockly_editor.js artifact from being generated by dashboard (via Sprockets) to being generated by code-studio (via browserify).  That makes the whole setup _more_ complicated for now, but hopefully _less_ complicated soon.
## TODO
- [x] Update chef recipe to create symlink in dashboard/public from code-studio to code-studio-package (like [this shared one](https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-apps/recipes/dashboard.rb#L48)) in managed environments.
- [x] Add package update to staging build
- (Out of scope) ~~Evaluate: How do we decide whether to build minified or not, and how do we tell dasboard which to look for?~~
- (Out of scope) ~~Get incremental builds working (in scope?)~~
- (Out of scope) ~~Get watchify working (in scope?)~~

Bjvanminnen commented on Rakefile:
```diff
@@ -303,6 +327,7 @@ namespace :install do
   tasks << :blockly_symlink
   tasks << :apps if CDO.build_apps
   tasks << :shared if CDO.build_shared_js
+  tasks << :code_studio
```
This shouldn't be gated on CDO.build_code_studio?

islemaster replied:
I thought about that (we do elsewhere) but I'm confused about why this is environment-specific.  The important step here is creating the correct symlink in dashboard/public.  Don't we need to do that in any environment?

Bjvanminnen replied:
That does seem accurate, but we have it conditioned for our other parts. Maybe we create the symlink by default elsewhere, and only need to run this install step if we're replacing it with a symlink to built code?

islemaster replied:
So, we do have a [chef recipe that creates these symlinks](https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-apps/recipes/dashboard.rb#L48) which I will need to update (adding a TODO to this PR description) but [the `local_environment?` check](https://github.com/code-dot-org/code-dot-org/blob/staging/Rakefile#L217) inside the `install:code_studio` task makes the install task a no-op in chef-configured environments anyway.  

For local development (or any non-chef environment) this is the only way I can find that we create those symlinks.  @wjordan any idea if I'm missing anything here?

islemaster replied:
[Commit 550b9ce](https://github.com/code-dot-org/code-dot-org/commit/550b9ce6d0bf47dba9a75631d89ae9fc52cc6e0f) gates this task on configuration and set appropriate defaults for different environments.

Bjvanminnen commented on code-studio/package.json:
```diff
@@ -0,0 +1,36 @@
+    "build:min": "browserify src/js/blockly_editor.js -t [ babelify --presets [es2015] ] | uglifyjs -o built/js/blockly_editor.js",
+    "prebuild": "mkdir -p built/js",
+    "prebuild:min": "mkdir -p built/js",
+    "clean": "rm -rf built",
```
so i don't have a strong opinion on built vs. build, but build might be better just because we have a precedent (in apps)

islemaster replied:
:grin: Just seeing if I can get away with it.  I remember being confused by the 'build' directory when I first encountered it.  Does it contain build output or build scripts, or some other build-critical thing?  Personally I find 'built' (the adjective) less confusing than 'build' (the verb or imperative).

islemaster replied:
I've renamed directory `built` => `build` to be consistent with other places in our repo.

Bjvanminnen commented on code-studio/src/js/blockly_editor.js:
```diff
@@ -0,0 +1,163 @@
+import "codemirror/mode/xml/xml";
+import "codemirror/mode/javascript/javascript";
+import "@cdo/vendor/js/codemirror.inline-attach";
+import "marked";
```
I thought we had decided to hold off on ES6 until first finishing the pipeline work.

islemaster replied:
We did!  My bad!  I'll put this back.

islemaster replied:
Done - no more ES6 here (yet).

islemaster commented:
Question for all: Am I jumping the gun creating a vendor package here, since the only thing it currently contains is `codemirror.inline-attach.js` which is only used by the code-studio package right now?

Alternative: Put `codemirror.inline-attach.js` in a vendor/js directory inside the code-studio package.  Don't add a root vendor package until we actually have a common vendor dependency to put there.  

Maybe we won't ever need to - it seems to me that dashboard is a client to apps, and it wouldn't be inappropriate for the code-studio package to have a declared dependency on the apps package, and its dependencies in turn.

islemaster commented:
@Bjvanminnen @joshlory @wjordan I think this thing is nearly ready to check in.  I could use some advice on what's possible to test here.

I'm mostly mirroring the shared package with regard to Rakefile stuff and what build commands run on staging.  I hope that works out for now - and makes it easy to clean them up together, later.

There's no special environment-specific stuff happening to the code-studio package right now, except for whatever Sprockets does with assets (which might be sufficient for now).  As Josh pointed out, this is more or less a nonissue for this PR since the blockly_editor.js asset is only used on levelbuilder.

I went ahead and removed the not-yet-needed vendor package and moved the one vendor script into a vendor directory in the code-studio package, which I symlink into node_modules as a post-install step to make requires easy.  This sounds hacky, but it's actually the recommended approach in [The
Browserify Handbook](https://github.com/substack/browserify-handbook#avoiding-).

wjordan commented:
I'm pretty sure the browserify handbook is obsolete with recent versions of npm. I don't think we need to be reaching into node_modules and managing symlinks manually. I'll get back to you on the more modern recommended approach to this using npm scoped packages and [linklocal](https://github.com/timoxley/linklocal) (which they may have bundled into npm@3, or they're still working on it). [**Edit**: they're still working on it.]

wjordan commented:
We're already using linklocal to manage package dependencies- for reference, see #2414 for the merged code, and see this [comment](https://github.com/code-dot-org/code-dot-org/pull/2376#discussion_r30151814) in #2376 for discussion/context.

islemaster commented:
I'm surprised to hear the handbook is obsolete, since it's still fairly prominent on browserify.org.  I'll update everything to use linklocal and post for review again.

islemaster commented:
@wjordan Okay, see latest change for using local paths and linklocal.  I love that, under the hood, this is exactly the same solution - it's just got some tooling to make it declarative instead of imperative.

islemaster commented:
Note: Merge conflicts here are due to #6096 revert of @Hamms authored hints work, which we've temporarily undone because we found a small regression.  Conflict should be easy to resolve once that change is re-submitted to staging.

Hamms commented:
Resubmitted! #6102
